src/live_restore/live_restore_fs.c

```diff
@@ -0,0 +1,1248 @@
+    if (ret == WT_NOTFOUND || !exist)
+        return (ENOENT);
+
+    /* If the file is the top layer, rename it and leave a tombstone behind. */
+    if (which == WT_LIVE_RESTORE_FS_LAYER_DESTINATION) {
+        WT_ERR(__live_restore_fs_filename(&lr_fs->destination, session, from, &path_from));
+        WT_ERR(__live_restore_fs_filename(&lr_fs->destination, session, to, &path_to));
+        WT_ERR(lr_fs->os_file_system->fs_rename(
+          lr_fs->os_file_system, wt_session, path_from, path_to, flags));
+        __wt_free(session, path_from);
+        __wt_free(session, path_to);
+
+        /* Create a tombstone for the file. */
+        WT_ERR(__live_restore_fs_create_tombstone(fs, session, to, flags));
+        /* Create a tombstone for the old file as well. */
+        WT_ERR(__live_restore_fs_create_tombstone(fs, session, from, flags));
+    }
+
+err:
+    __wt_free(session, path_from);
+    __wt_free(session, path_to);
+    return (ret);
+}
+
+/*
+ * __live_restore_fs_size --
+ *     Get the size of a file in bytes, by file name.
+ */
+static int
+__live_restore_fs_size(
+  WT_FILE_SYSTEM *fs, WT_SESSION *wt_session, const char *name, wt_off_t *sizep)
+{
+    WT_DECL_RET;
+    WT_LIVE_RESTORE_FS *lr_fs;
+    WT_LIVE_RESTORE_FS_LAYER_TYPE which;
+    WT_SESSION_IMPL *session;
+    char *path;
+    bool exist;
+
+    session = (WT_SESSION_IMPL *)wt_session;
+    lr_fs = (WT_LIVE_RESTORE_FS *)fs;
+
+    exist = false;
+    path = NULL;
+
+    WT_RET_NOTFOUND_OK(__live_restore_fs_find_layer(fs, session, name, &which, &exist));
+    if (ret == WT_NOTFOUND || !exist)
+        return (ENOENT);
+
+    /* The file will always exist in the destination. This the is authoritative file size. */
+    WT_ASSERT(session, which == WT_LIVE_RESTORE_FS_LAYER_DESTINATION);
+    WT_RET(__live_restore_fs_filename(&lr_fs->destination, session, name, &path));
+    ret = lr_fs->os_file_system->fs_size(lr_fs->os_file_system, wt_session, path, sizep);
+
+    __wt_free(session, path);
+
+    return (ret);
+}
+
+/*
+ * __live_restore_fs_terminate --
+ *     Terminate the file system.
+ */
+static int
+__live_restore_fs_terminate(WT_FILE_SYSTEM *fs, WT_SESSION *wt_session)
+{
+    WT_LIVE_RESTORE_FS *lr_fs;
+    WT_SESSION_IMPL *session;
+
+    session = (WT_SESSION_IMPL *)wt_session;
+    lr_fs = (WT_LIVE_RESTORE_FS *)fs;
+
+    WT_ASSERT(session, lr_fs->os_file_system != NULL);
+    WT_RET(lr_fs->os_file_system->terminate(lr_fs->os_file_system, wt_session));
+
+    __wt_free(session, lr_fs->source.home);
+    /* TODO: Do we free ourselves here? */
+    return (0);
+}
+
+/*
+ * __wt_os_live_restore_fs --
+ *     Initialize a live restore file system configuration.
+ */
+int
+__wt_os_live_restore_fs(WT_SESSION_IMPL *session, WT_CONFIG_ITEM *source_cfg,
+  const char *destination, WT_FILE_SYSTEM **fsp)
+{
+    WT_LIVE_RESTORE_FS *lr_fs;
+
+    WT_RET(__wt_calloc_one(session, &lr_fs));
+    WT_RET(__wt_os_posix(session, &lr_fs->os_file_system));
```

Comment:
nit: We should WT_ERR and free lr_fs on failure.
Perhaps it's not necessary since we're almost guaranteed to crash when we can't init the file system.

Reply:
[3f680da](https://github.com/wiredtiger/wiredtiger/pull/11272/commits/3f680da7eafab482a6f15265f1015c817edfa1d5)